pil-analyzer/src/traits_resolver.rs

```diff
@@ -0,0 +1,131 @@
+
+        let resolve_references = |expr: &Expression| {
+            expr.all_children()
+                .filter_map(|expr| {
+                    if let Expression::Reference(
+                        _,
+                        Reference::Poly(
+                            reference @ PolynomialReference {
+                                type_args: Some(_), ..
+                            },
+                        ),
+                    ) = expr
+                    {
+                        self.resolve_reference_trait(reference, trait_impls)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>()
+        };
+
+        for (_, (_, def)) in self.definitions.iter() {
+            if let Some(FunctionValueDefinition::Expression(TypedExpression { e: expr, .. })) = def
+            {
+                for (key, value) in resolve_references(expr) {
+                    result
+                        .entry(key.0)
+                        .or_insert_with(HashMap::new)
+                        .insert(key.1, value);
+                }
+            }
+        }
+
+        for identity in self.identities.iter() {
+            for selector in identity
+                .left
+                .selector
+                .iter()
+                .chain(identity.right.selector.iter())
+            {
+                for (key, value) in resolve_references(selector) {
+                    result
+                        .entry(key.0)
+                        .or_insert_with(HashMap::new)
+                        .insert(key.1, value);
+                }
+            }
+            for (key, value) in resolve_references(identity.left.expressions.as_ref()) {
+                result
+                    .entry(key.0)
+                    .or_insert_with(HashMap::new)
+                    .insert(key.1, value);
+            }
+            for (key, value) in resolve_references(identity.right.expressions.as_ref()) {
+                result
+                    .entry(key.0)
+                    .or_insert_with(HashMap::new)
+                    .insert(key.1, value);
+            }
+        }
+
+        result
+    }
+    fn resolve_reference_trait(
+        &self,
+        reference: &PolynomialReference,
+        trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
+    ) -> Option<SolvedImpl> {
+        let name = reference.name.clone();
+        let (trait_decl_name, trait_fn_name) = name.rsplit_once("::")?;
+        if let Some(impls) = trait_impls.get(trait_decl_name) {
+            for impl_ in impls.iter() {
+                let Type::Tuple(TupleType { ref items }) = impl_.type_scheme.ty else {
+                    unreachable!()
+                };
+                let type_args = reference.type_args.as_ref().unwrap().to_vec();
+
+                let expr = impl_.function_by_name(trait_fn_name).unwrap();
+                if type_args == *items {
```

Comment:
You need to call unify or are you saying this is fine because trait impls are not generic yet? If yes, please add a todo